File: src/979. Distribute Coins in Binary Tree/main.go
// Solved by @kitanoyoru

package main


type TreeNode struct {
  Val int
  Left *TreeNode
  Right *TreeNode
}

func abs(x int) int {
  if x < 0 {
    return -x
  }
  return x
}

func distributeCoins(root *TreeNode) int {
  ans := 0

  if root.Left != nil {
    ans += distributeCoins(root.Left)
    root.Val += root.Left.Val - 1
    ans += abs(root.Left.Val - 1)
  }
  if root.Right != nil {
    ans += distributeCoins(root.Right)
    root.Val += root.Right.Val - 1
    ans += abs(root.Right.Val - 1)
  }

  return ans
}

File: src/979. Distribute Coins in Binary Tree/main.c
// Solved by @kitanoyoru

struct TreeNode {
  int val;
  struct TreeNode *left;
  struct TreeNode *right;
};

int distributeCoins(struct TreeNode* root){
  int ans = 0;

  if (root->left != NULL) {
    ans += distributeCoins(root->left);
    root->val += root->left->val - 1;
    ans += abs(root->left->val - 1);
  }
  if (root->right != NULL) {
    ans += distributeCoins(root->right);
    root->val += root->right->val - 1;
    ans += abs(root->right->val - 1);
  }

  return ans;
}
